common/browser_process.go

```diff
@@ -186,37 +181,48 @@ func execute(
 		}
 	}()
 
-	return cmd, stdout, done, nil
+	return cmd, done, nil
 }
 
-// parseDevToolsURL grabs the websocket address from chrome's output and returns it.
-func parseDevToolsURL(ctx context.Context, rc io.Reader) (wsURL string, _ error) {
-	type result struct {
-		devToolsURL string
-		err         error
-	}
-	c := make(chan result, 1)
-	go func() {
-		const prefix = "DevTools listening on "
-
-		scanner := bufio.NewScanner(rc)
-		for scanner.Scan() {
-			if s := scanner.Text(); strings.HasPrefix(s, prefix) {
-				c <- result{
-					strings.TrimPrefix(strings.TrimSpace(s), prefix),
-					nil,
-				}
-				return
-			}
+// getDevToolsURL returns the DevTools WebSocket address by reading the
+// DevToolsActivePort file in the data directory.
+func getDevToolsURL(dataDir string) (wsURL string, rerr error) {
+	var (
+		f                  *os.File
+		fpath              = filepath.Join(dataDir, "DevToolsActivePort")
+		maxReadAttempts    = 10
+		readAttemptDelayMS = 50
+	)
+
+	// The browser might not have created the file yet, so try reading it
+	// multiple times after a slight delay.
+	for readAttempts := 0; readAttempts < maxReadAttempts; readAttempts++ {
+		var err error
+		f, err = os.Open(fpath) //nolint:gosec  // false positive, https://github.com/securego/gosec/issues/439
+		if errors.Is(err, os.ErrNotExist) {
+			time.Sleep(time.Duration(readAttemptDelayMS) * time.Millisecond)
```

Comment:
Might we need to use a select statement here with `time.Timer` and also watch for the `context` being cancelled?